jabgui/src/main/java/org/jabref/gui/importer/BookCoverFetcher.java

```diff
@@ -0,0 +1,150 @@
+package org.jabref.gui.importer;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jabref.gui.externalfiletype.ExternalFileType;
+import org.jabref.gui.externalfiletype.ExternalFileTypes;
+import org.jabref.gui.frame.ExternalApplicationsPreferences;
+import org.jabref.logic.FilePreferences;
+import org.jabref.logic.importer.FetcherException;
+import org.jabref.logic.net.URLDownload;
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+import org.jabref.model.entry.identifier.ISBN;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides functions for downloading book covers for new entries.
+ */
+public class BookCoverFetcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BookCoverFetcher.class);
+
+    private static final Pattern URL_JSON_PATTERN = Pattern.compile("^\\s*\\{\\s*\"url\"\\s*:\\s*\"([^\"]*)\"\\s*\\}\\s*$");
+
+    private static final String URL_FETCHER_URL = "https://bookcover.longitood.com/bookcover/";
+    private static final String IMAGE_FALLBACK_URL = "https://covers.openlibrary.org/b/isbn/";
+    private static final String IMAGE_FALLBACK_SUFFIX = "-L.jpg";
+
+    public static Optional<BibEntry> withAttachedCoverFileIfExists(Optional<BibEntry> possible, BibDatabaseContext databaseContext, FilePreferences filePreferences, ExternalApplicationsPreferences externalApplicationsPreferences) {
```

Comment:
Optional handling is not OK here.

Methods may return optional, but not get optional passed.

You see it here, because you skip all logic if Optional is empty.

The trick is

    return fetcher.performSearchById(text).map(bibEntry -> withCoversAttached(bibEntry) ...)

Then, the method is only called if BibEntry exists - and an empty Optional is returned if no search result is there.